internal/os_updater/emt.go

```diff
@@ -6,26 +6,314 @@
-type EMTDownloader struct{}
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/intel/intel-inb-manageability/internal/inbd/utils"
+	pb "github.com/intel/intel-inb-manageability/pkg/api/inbd/v1"
+	"github.com/spf13/afero"
+	"golang.org/x/sys/unix"
+)
+
+var (
+	configFilePath = "/etc/intel_manageability.conf"
+	JWTTokenPath   = "/etc/intel_edge_node/tokens/release-service/access_token"
+	downloadDir    = "/var/cache/manageability/repository-tool/sota"
+	// OsUpdateTool will be changed in 3.1 release. Have to change the name and API call.
+	// Check https://github.com/intel-sandbox/os.linux.tiberos.ab-update.go/blob/main/README.md
+	osUpdateToolPath = "/usr/bin/os-update-tool.sh"
+)
+
+// EMTDownloader is the concrete implementation of the IDownloader interface
+// for the Emt OS.
+type EMTDownloader struct {
+	request          *pb.UpdateSystemSoftwareRequest
+	readJWTTokenFunc func() (string, error)
+	statfs           func(path string, stat *unix.Statfs_t) error
+	httpClient       *http.Client
+	requestCreator   func(method string, url string, body io.Reader) (*http.Request, error)
+	fs               afero.Fs
+}
+
+// NewEMTDownloader creates a new EMTDownloader.
+func NewEMTDownloader(request *pb.UpdateSystemSoftwareRequest) *EMTDownloader {
+	return &EMTDownloader{
+		request:          request,
+		readJWTTokenFunc: defaultReadJWTToken,
+		statfs:           unix.Statfs,
+		httpClient:       &http.Client{},
+		requestCreator:   http.NewRequest,
+		fs:               afero.NewOsFs(),
+	}
+}
+
+func defaultReadJWTToken() (string, error) {
+	file, err := os.Open(JWTTokenPath)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	token, err := os.ReadFile(JWTTokenPath)
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(string(token)), nil
+}
 
 // Download implements IDownloader.
-func (u *EMTDownloader) Download() error {	
-	panic("unimplemented")
+func (t *EMTDownloader) Download() error {
+	config, err := LoadConfig(configFilePath)
+	if err != nil {
+		fmt.Println("Error loading intel_manageability.conf:", err)
+		return err
+	}
+
+	// Perform source verification
+	if !IsTrustedRepository(t.request.Url, config) {
+		errMsg := fmt.Sprintf("URL '%s' is not in the list of trusted repositories.", t.request.Url)
+		fmt.Println(errMsg)
+		return errors.New(errMsg)
+	}
+
+	fmt.Println("Downloading update from", t.request.Url)
+
+	// Check available space on disk
+	isDiskEnough, err := t.checkDiskSpace()
+	if err != nil {
+		fmt.Println("Error checking disk space:", err)
+		return errors.New(err.Error())
+	}
+
+	if !isDiskEnough {
+		errMsg := "Insufficient disk space."
+		fmt.Println(errMsg)
+		return errors.New(errMsg)
+	}
+
+	fmt.Println("Disk space enough. Proceeding to download the artifact.")
+
+	// Download file
+	err = t.downloadFile()
+	if err != nil {
+		fmt.Println("Error downloading the file:", err)
+		return err
+	}
+
```

Comment:
should SHA be checked here before passing to update tool?

Reply:
Added SHA verification before running UT command.